File: blue-core/src/blue/soundObject/Note.java
/*
 * blue - object composition environment for csound Copyright (c) 2001-2003
 * Steven Yi (stevenyi@gmail.com)
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB. If not, write to the Free
 * Software Foundation Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307
 * USA
 */

package blue.soundObject;

import java.io.Serializable;
import java.util.ArrayList;

import org.apache.commons.lang.text.StrBuilder;

import blue.BlueSystem;
import blue.utility.NumberUtilities;
import blue.utility.ObjectUtilities;
import blue.utility.ScoreExpressionParser;

/**
 * Title: blue (Object Composition Environment) Description: Copyright:
 * Copyright (c) steven yi Company: steven yi music
 * 
 * @author steven yi
 * @version 1.0
 */

public class Note implements Serializable, Comparable {

    private String[] fields;

    private float subjectiveDuration;

    boolean isTied = false;

    private Note() {
        /*
         * try { noteInit(input); } catch(Exception e) { e.printStackTrace(); }
         */
    }

    public static Note createNote(int numFields) {
        Note n = new Note();

        n.fields = new String[numFields];

        for (int i = 0; i < numFields; i++) {
            n.setPField(Integer.toString(i), i + 1);
        }

        return n;
    }

    public static Note createNote(String input) throws NoteParseException {
        return createNote(input, null);
    }

    public static Note createNote(String input, Note previousNote)
            throws NoteParseException {
        if (!input.startsWith("i")) {
            String errorMessage = BlueSystem.getString("note.invalidNoteText")
                    + " " + input;

            throw new NoteParseException(errorMessage, input);
        }

        try {
            Note n = new Note();
            final String cleanNoteString = input
                    .substring(input.indexOf("i") + 1);
            n.noteInit(cleanNoteString, previousNote);
            return n;
        } catch (Exception e) {
            String errorMessage = BlueSystem.getString("note.invalidNoteText")
                    + " " + input;
            throw new NoteParseException(errorMessage, input);
        }
    }

    private void noteInit(String input, Note previousNote) throws Exception {
        // If for any reason there should be an exception
        // let it bubble up and the note factory method will
        // return null

        int i = 0;
        ArrayList buffer = new ArrayList();
        int size = input.length();

        int start = 0;

        // the following code might be better off using
        // regular expressions, but should wait until
        // Java 1.4 is available on all platforms

        // PARSES PFIELDS FROM STRING
        while (i < size) {
            if (input.charAt(i) == '\"') {
                start = i++;

                while (i < size && input.charAt(i) != '\"') {
                    i++;
                }

                buffer.add(input.substring(start, ++i));
                // i++;
            } else if (input.charAt(i) == '[') {
                start = ++i;
                while (i < size && input.charAt(i) != ']') {
                    i++;
                }

                float val = ScoreExpressionParser.eval(input
                        .substring(start, i));

                i++;

                buffer.add(Float.toString(val));
            } else if (input.charAt(i) == ' ' || input.charAt(i) == '\t') {
                while (i < size
                        && (input.charAt(i) == ' ' || input.charAt(i) == '\t')) {
                    i++; // eat up empty spaces or tabs
                }
            } else {
                start = i;
                while (i < size
                        && (input.charAt(i) != ' ' && input.charAt(i) != '\t')) {
                    i++;
                }
                buffer.add(input.substring(start, i));
            }
        }

        

        if(previousNote != null) {
            boolean performCarry = buffer.get(0).equals(previousNote.getPField(1));

            if(!performCarry) {
                try {
                    int instr1 = (int) (Float.parseFloat((String)buffer.get(0)));
                    int instr2 = (int) (Float.parseFloat(previousNote.getPField(1)));

                    if(instr1 == instr2) {
                        performCarry = true;
                    }
                } catch (NumberFormatException nfe) {
                    performCarry = false;
                }
            }


            if(performCarry) {
                int numFieldsToCopy = previousNote.getPCount() - buffer.size();
                if(numFieldsToCopy > 0) {
                    for(i = previousNote.getPCount() - numFieldsToCopy; i < previousNote.getPCount(); i++) {
                        buffer.add(previousNote.getPField(i + 1));
                    }
                }
            }
        }
        
        // INITIALIZES PFIELD ARRAY
        fields = (String[]) buffer.toArray(new String[buffer.size()]);

        if(previousNote != null) {


            // SWAP PERIODS WITH VALUE FROM PREVIOUS NOTE
            for (i = 0; i < fields.length; i++) {
                if (fields[i].equals(".")) {
                    fields[i] = previousNote.getPField(i + 1);
                }
            }
        }

        float dur = Float.parseFloat(fields[2]);

        setSubjectiveDuration(dur);
        setTied(dur < 0.0f);
    }

    public String toString() {
        StrBuilder temp = new StrBuilder();

        int strSize = 1;

        for (int i = 0; i < fields.length; i++) {
            strSize += fields[i].length() + 1;
        }

        temp.ensureCapacity(strSize);

        temp.append("i");

        int size = fields.length;
        for (int i = 0; i < size; i++) {
            if (i == 2) {
                if (this.isTied) {
                    temp.append("-");
                }
                temp.append(NumberUtilities.formatFloat(subjectiveDuration))
                        .append("\t");
            } else {
                temp.append(fields[i]);

                if (i < size - 1) {
                    temp.append("\t");
                }
            }
        }

        return temp.toString();
    }

    public void setStartTime(float yo) {
        fields[1] = Float.toString(yo);
    }

    public float getStartTime() {
        return Float.parseFloat(fields[1]);
    }

    public void setSubjectiveDuration(float dur) {
        subjectiveDuration = Math.abs(dur);
    }

    public float getSubjectiveDuration() {
        return subjectiveDuration;
    }

    public float getObjectiveDuration() {
        return getSubjectiveDuration();
    }

    public int getPCount() {
        return fields.length;
    }

    public String getPField(int i) {
        return fields[i - 1];
    }

    public void setPField(String arg, int index) {
        if (index == 3) {
            setSubjectiveDuration(Float.parseFloat(arg));
        }
        fields[index - 1] = arg;
    }

    public Object clone() {
        return ObjectUtilities.clone(this);
    }

    public int compareTo(Object pObj) {
        Note a = (Note) pObj;

        float t1 = this.getStartTime();
        float t2 = a.getStartTime();

        if (t1 > t2) {
            return 1;
        } else if (t1 < t2) {
            return -1;
        }

        return 0;
    }

    public static void main(String[] args) {
        Note testNote;
        try {
            testNote = Note.createNote("i1 2 3 4 5");
            testNote.setSubjectiveDuration(12f);
            System.out.println(testNote.toString());

            testNote = Note.createNote("i1 [4 + 23] 5 4");
            testNote.setSubjectiveDuration(12f);
            System.out.println(testNote.toString());

            testNote = Note.createNote("i1 [4 + 23 -] 5 4");
            testNote.setSubjectiveDuration(12f);
            System.out.println(testNote.toString());

        } catch (NoteParseException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    /**
     * @return Returns the isTied.
     */
    public boolean isTied() {
        return isTied;
    }

    /**
     * @param isTied
     *            The isTied to set.
     */
    public void setTied(boolean isTied) {
        this.isTied = isTied;
    }

    /**
     * @return
     */
    public float getEndTime() {
        return getStartTime() + getSubjectiveDuration();
    }
}

File: packages/blue-data/src/sound-objects/note.ts
import { formatBlueNumber, formatJavaDouble } from '../utilities/number-format';

const TOKEN_PATTERN = /"[^"]*"|\[[^\]]*\]|\S*/g;

function evalBracketExpression(expr: string): string {
  try {
    const sanitized = expr
      .replace(/[^0-9+\-*/().eE\s]/g, '')
      .trim();
    if (!sanitized) return '0';
    const result = new Function(`return (${sanitized})`)();
    if (typeof result === 'number' && isFinite(result)) {
      return result.toString();
    }
    return expr;
  } catch {
    return expr;
  }
}

export class Note {
  private _pFields = new Map<number, string>();
  private _startTime = 0;
  private _subjectiveDuration = 0;
  isTied = false;

  static createNote(numPFields: number): Note {
    const note = new Note();
    for (let i = 1; i <= numPFields; i++) {
      note._pFields.set(i, (i - 1).toString());
    }
    return note;
  }

  static createBlank(numPFields: number): Note {
    return Note.createNote(numPFields);
  }

  static fromOther(other: Note): Note {
    const note = new Note();
    note._startTime = other._startTime;
    note._subjectiveDuration = other._subjectiveDuration;
    note.isTied = other.isTied;
    for (const [k, v] of other._pFields) {
      note._pFields.set(k, v);
    }
    return note;
  }

  static createNoteFromText(text: string, previousNote?: Note | null): Note | null {
    const trimmed = text.trim();
    if (!trimmed.startsWith('i')) return null;

    try {
      const n = new Note();
      const cleanText = trimmed.substring(trimmed.indexOf('i') + 1);
      n.noteInit(cleanText, previousNote ?? null);
      return n;
    } catch {
      return null;
    }
  }

  private noteInit(input: string, previousNote: Note | null): void {
    const buffer: string[] = [];

    const m = input.match(TOKEN_PATTERN);
    if (m) {
      for (const str of m) {
        if (!str) continue;
        if (str.charAt(0) === '[') {
          const evaluated = evalBracketExpression(str.substring(1, str.length - 1));
          buffer.push(evaluated);
        } else {
          buffer.push(str);
        }
      }
    }

    if (buffer.length < 3) throw new Error('Insufficient pfields');

    if (previousNote !== null) {
      let performCarry = buffer[0] === previousNote.getPField(1);

      if (!performCarry) {
        try {
          const instr1 = parseInt(buffer[0], 10);
          const instr2 = parseInt(previousNote.getPField(1) ?? '', 10);
          if (instr1 === instr2) performCarry = true;
        } catch { /* not numeric */ }
      }

      if (performCarry) {
        const numFieldsToCopy = previousNote.getPCount() - buffer.length;
        if (numFieldsToCopy > 0) {
          for (let i = previousNote.getPCount() - numFieldsToCopy; i < previousNote.getPCount(); i++) {
            buffer.push(previousNote.getPField(i + 1) ?? '');
          }
        }
      }
    }

    for (let i = 0; i < buffer.length; i++) {
      this._pFields.set(i + 1, buffer[i]);
    }

    if (previousNote !== null) {
      for (let i = 0; i < buffer.length; i++) {
        if (buffer[i] === '.') {
          const prevVal = previousNote.getPField(i + 1);
          if (prevVal !== undefined) {
            this._pFields.set(i + 1, prevVal);
          }
        }
      }
    }

    const startTimeStr = this._pFields.get(2) ?? '0';
    const durStr = this._pFields.get(3) ?? '0';
    const rawDur = parseFloat(durStr);
    this._startTime = parseFloat(startTimeStr);
    this.setSubjectiveDuration(rawDur);
    this.setTied(rawDur < 0);
  }

  getStartTime(): number { return this._startTime; }
  setStartTime(time: number): void {
    this._startTime = time;
    this._pFields.set(2, formatJavaDouble(time));
  }

  getSubjectiveDuration(): number { return this._subjectiveDuration; }
  setSubjectiveDuration(duration: number): void {
    this._subjectiveDuration = Math.abs(duration);
  }

  getObjectiveDuration(): number { return this.getSubjectiveDuration(); }

  getEndTime(): number { return this._startTime + this._subjectiveDuration; }

  isTiedNote(): boolean { return this.isTied; }
  setTied(tied: boolean): void { this.isTied = tied; }

  getPCount(): number { return this._pFields.size; }

  getPField(index: number): string | undefined {
    return this._pFields.get(index);
  }

  setPField(value: string, index: number): void {
    this._pFields.set(index, value);
    if (index === 2) {
      this._startTime = parseFloat(value);
    }
    if (index === 3) {
      this.setSubjectiveDuration(parseFloat(value));
    }
  }

  getPFields(): Map<number, string> {
    return new Map(this._pFields);
  }

  toScoreText(): string {
    const parts: string[] = [];
    parts.push('i' + (this._pFields.get(1) ?? '0'));
    parts.push(formatJavaDouble(this._startTime));
    if (this.isTied) {
      parts.push('-' + formatBlueNumber(this._subjectiveDuration));
    } else {
      parts.push(formatBlueNumber(this._subjectiveDuration));
    }
    for (let i = 4; ; i++) {
      const val = this._pFields.get(i);
      if (val === undefined) break;
      parts.push(val);
    }
    return parts.join('\t');
  }

  toString(): string {
    return this.toScoreText();
  }

  deepCopy(): Note {
    return Note.fromOther(this);
  }
}
